core/quivr_core/rag/quivr_rag_langgraph.py

```diff
@@ -0,0 +1,930 @@
+        " If the final answer to the user is 'I don't know' or 'I don't have enough information' or 'I'm not sure', "
+        " this variable should be 'false'"
+    )
+
+
+class UpdatedPromptAndTools(BaseModel):
+    prompt_reasoning: Optional[str] = Field(
+        default=None,
+        description="The step-by-step reasoning that leads to the updated system prompt",
+    )
+    prompt: Optional[str] = Field(default=None, description="The updated system prompt")
+
+    tools_reasoning: Optional[str] = Field(
+        default=None,
+        description="The reasoning that leads to activating and deactivating the tools",
+    )
+    tools_to_activate: Optional[List[str]] = Field(
+        default=None, description="The list of tools to activate"
+    )
+    tools_to_deactivate: Optional[List[str]] = Field(
+        default=None, description="The list of tools to deactivate"
+    )
+
+
+class AgentState(TypedDict):
+    # The add_messages function defines how an update should be processed
+    # Default is to replace. add_messages says "append"
+    messages: Annotated[Sequence[BaseMessage], add_messages]
+    reasoning: List[str]
+    chat_history: ChatHistory
+    docs: list[Document]
+    files: str
+    tasks: List[str]
+    instructions: str
+    tool: str
+
+
+class InstructionsState(TypedDict):
+    instructions: str
+
+
+class QuestionsState(TypedDict):
+    individual_chat_history: ChatHistory
+    question: str
+
+
+class GenerateInputState(AgentState, QuestionsState):
+    pass
+
+
+class IdempotentCompressor(BaseDocumentCompressor):
+    def compress_documents(
+        self,
+        documents: Sequence[Document],
+        query: str,
+        callbacks: Optional[Callbacks] = None,
+    ) -> Sequence[Document]:
+        """
+        A no-op document compressor that simply returns the documents it is given.
+
+        This is a placeholder until a more sophisticated document compression
+        algorithm is implemented.
+        """
+        return documents
+
+
+class QuivrQARAGLangGraph:
+    def __init__(
+        self,
+        *,
+        retrieval_config: RetrievalConfig,
+        llm: LLMEndpoint,
+        vector_store: VectorStore | None = None,
+    ):
+        """
+        Construct a QuivrQARAGLangGraph object.
+
+        Args:
+            retrieval_config (RetrievalConfig): The configuration for the RAG model.
+            llm (LLMEndpoint): The LLM to use for generating text.
+            vector_store (VectorStore): The vector store to use for storing and retrieving documents.
+            reranker (BaseDocumentCompressor | None): The document compressor to use for re-ranking documents. Defaults to IdempotentCompressor if not provided.
+        """
+        self.retrieval_config = retrieval_config
+        self.vector_store = vector_store
+        self.llm_endpoint = llm
+
+        self.graph = None
+
+    def get_reranker(self, **kwargs):
```

Comment:
the reason is that in https://github.com/QuivrHQ/quivr/pull/3424/files#diff-4b557c1696776056776ecb2a2beb9d21b4ce9d6a0159e8464f53dc88d820ee19R615-R622 we need to be able to modify the original configuration "on-the-fly", so that we need to keep the initial configuration (stored in `reranker_config`), while also instantiating a reranker with parameters defined inside the workflow.

This is a more generic need, i.e. having a "parent" configuration (input from the user or from a database), and some "child" configurations which are modified on-the-fly by the system. This is a simple/quick solution before imagining a more generic one

Reply:
I see the need for this at runtime my comments was only about the dict usage kwargs instead of passing structs.